client/starwhale/cli/assistance/host.py

```diff
@@ -0,0 +1,109 @@
+        with subprocess.Popen(
+            self.args,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+        ) as p:
+            assert p.stdin is not None
+            stdin_buffer = ChunkBuffer(256 * 1024, 64)
+            FileWriter(p.stdin, stdin_buffer).start()
+            BrokerReader(f"{self.broker_url}/file/stdin", stdin_buffer).start()
+            assert p.stdout is not None
+            stdout_buffer = ChunkBuffer(256 * 1024, 64)
+            FileReader(p.stdout, stdout_buffer).start()
+            stdout_writer = BrokerWriter(
+                f"{self.broker_url}/file/stdout", stdout_buffer
+            )
+            stdout_writer.start()
+            assert p.stderr is not None
+            stderr_buffer = ChunkBuffer(256 * 1024, 64)
+            FileReader(p.stderr, stderr_buffer).start()
+            stderr_writer = BrokerWriter(
+                f"{self.broker_url}/file/stderr", stderr_buffer
+            )
+            stderr_writer.start()
+            while not self.stopped:
+                try:
+                    exit_code = p.wait(0.5)
+                    break
+                except subprocess.TimeoutExpired:
+                    pass
+            if self.stopped:
+                p.terminate()
+                exit_code = p.wait()
+            self.run_until_success(lambda: self.update_exit_code(exit_code))
+            stdout_writer.join()
+            stderr_writer.join()
+
+    def keep_alive(self) -> None:
+        requests.post(f"{self.broker_url}/keepAlive?end=host")
+
+    def update_exit_code(self, exit_code: int) -> None:
+        resp = requests.post(f"{self.broker_url}?exit_code={exit_code}")
+        check_status_code(resp)
+        return resp.json()  # type: ignore
+
+
+class CommandRetriever(TaskRunner):
+    def __init__(self, broker_url: str):
+        TaskRunner.__init__(self)
+        self.broker_url = broker_url
+        self.executions: Dict[str, CommandExecutor] = {}
+
+    def run(self) -> None:
+        try:
+            while not self.stopped:
+                commands = self.run_until_success(self.retrieve_commands)
+                for command in commands:
+                    command_id = command["command_id"]
+                    args = command["args"]
+                    if command_id not in self.executions:
+                        print(f"run command {command_id}:", args)
+                        executor = CommandExecutor(
+                            f"{self.broker_url}/command/{command_id}", args
+                        )
+                        executor.start()
+                        self.executions[command_id] = executor
+                time.sleep(1)
+        finally:
+            for executor in self.executions.values():
+                executor.stop()
+                executor.join()
+
+    def retrieve_commands(self) -> List[Dict[str, Any]]:
+        resp = requests.get(f"{self.broker_url}/command")
```

Comment:
add `timeout` for requests.get?   requests http request use the infinite value as the default timeout.